Add support for OdcmObject Deprecation

mikepizzo opened the pull request:
Add ability to read Deprecation information from Revisions annotation.

MIchaelMainer commented:
Adding @zengin

baywet reviewed (commented):
@mikepizzo QQ: shouldn't a deprecation on an entity type/complex type also propagate to the navigation properties/collections/functions that use that type?

mikepizzo commented:
@MIchaelMainer -- absolutely; this PR was intended to drive that discussion. I also have a [draft PR](https://github.com/microsoftgraph/MSGraph-SDK-Code-Generator/pull/283) against the SDK showing how the templates might use this information.

@baywet -- how far the deprecation should propagate is a really good question.  I started minimally, to avoid clutter, under the assumption that, if a collection, property, or operation referenced a deprecated type then somewhere along the line the developer would reference that type and still get the obsolete warning during compile. But, I am still very nascent in my understanding of the way developers use the SDK, so I don't know which of those naturally apply.

baywet commented:
The service library would show a warning in that case but not necessarily the consumer.
Example: dotnet, somebody using a navigation property of which the type is deprecated. (Simple get scenario on the parent entity).
If they affect the property to a variable using var, they'll never get a warning if we don't tag the property.
This is why I think either viper needs to propagate things or we need to make sure API owners add the annotation to every reference of the deprecated type. I think the later is more error prone.

mikepizzo commented:
@baywet  - totally agree; if it makes sense for the deprecation to propagate I'd rather do it in Vipr and not rely on the schema owners getting it right.  I'll play around with how that would look, both in Vipr and in the Templates.

MIchaelMainer commented:
Can revisions apply to any CSDL element? I assume so since there wasn't an Applies to attribute defined for the term.

mikepizzo commented:
@baywet - see latest commit.  I added logic for propagating the deprecation to derived types as well as usages in properties, nav props, entity sets and singletons.

Also removed code in related PR from codegen templates not to try and check derived types and usages there, since it is not handled in vipr.  Helped clean up/simplify logic in the templates.

zengin commented:
@mikepizzo, we were discussing this with @MIchaelMainer last week and wondering whether the propagation of deprecation can happen at the metadata level. The reason is that metadata is the source of truth in majority of our scenarios, and it will be hard to track this implicit logic by looking at the metadata when we want to discover if a type is getting deprecated.

For example, if a property is about to get deprecated because its type's great great grandparent type is getting deprecated, it will be hard to see that parity between the SDK and the metadata.

Even if it is hard for service teams to get the propagation right, wouldn't it be better to have explicit annotation in metadata? How feasible would it be to propagate annotations via transformations in metadata itself? I think increased metadata size with redundant information would be one concern in this case.

Wondering what your thoughts regarding this are, thanks!

baywet commented:
@mikepizzo thanks for adding this! I'd add it to the functions for which the return type is a deprecated type as well. And I'm temped to say actions/functions that take a deprecated type as parameter too but we can debate on this one.

MIchaelMainer commented on src/Core/Vipr.Core/CodeModel/OdcmDeprecation.cs:
```diff
@@ -0,0 +1,21 @@
+        /// <summary>
+        /// String for Deprecation Version
+        /// </summary>
+        public string Version { get; set; }
```
RemovalDate would be helpful to customers to add to /// comments so that it gets propagated to intellisense - it would help them plan to make their change. I'd like this information to be added to the obsolete message in the templates

mikepizzo replied:
See above. I would prefer to have standard guidance that workloads include this in the deprecation message, rather than try to artificially augment the messages from the workloads.

MIchaelMainer commented:
Functions and actions should be considered deprecated if:
- bound to a deprecated type,
- returning a deprecated type

Deprecated parameters cause us an issue because renamed or removed parameters and their types cause signature changes for .NET and Java.

mikepizzo commented on src/Readers/Vipr.Reader.OData.v4/OdcmReader.cs:
```diff
@@ -785,6 +787,75 @@ private OdcmNamespace ResolveNamespace(string @namespace)
+                            if (isDeprecated)
+                            {
+                                odcmObject.Deprecation = new OdcmDeprecation { 
+                                    Description = message,
```
I had played around with constructing a message that included the element, deprecation date, and removal date, but these elements (where known) are likely to already be part of the message from the workload, and when I manufactured a message using this information it ended up just being redundant.

The removal date (where known) is useful for machine consumption (i.e., generating a Sunset header), but I ultimately decided to leave the contents of the message up to the workloads, and not try to artificially augment it.

We could always provide guidance (i.e., through API review) on the format (and contents) of the deprecation messages provided by workloads in the CSDL.

Thoughts?

MIchaelMainer replied:
I was incomplete in my comment. The reason why I'd like to see removal date and deprecation date is that we could add it to the generated /// comments which would become available in intellisense giving customers a heads-up when they use the API about our plans to deprecate the API.

mikepizzo replied:
yes, but how would we merge the information from the message with deprecation and (if supplied) removal date?  As I say, I tried to do this in an early implementation, but it ended up being redundant with the information in the message.  So, when generating the Obsolete attribute for deprecated elements in the SDK, I ended up just using the message from the deprecation annotation (which generally includes the deprecation and, if known, removal date).

it's easy enough to add the additional properties to the Deprecation object (either now, or later) -- I just didn't want to add them until we had a scenario where we knew someone was going to use them.

let me know either way; I can add them now if you like, or we can wait for a use case.

mikepizzo commented:
@MIchaelMainer /@baywet -- I was playing around with adding deprecation to functions and methods based on binding/return type.  However, it occurred to me that, if the binding type is deprecated, they will already have to create an instance of that deprecated type to call it, and if the return type is deprecated then they will already have to create a variable of the return type in order to consume it.  So perhaps marking the action/function as deprecated is redundant?

It's easy to do, and I'm happy to add the code, I just wanted to be sure it wasn't overkill.

Thoughts?

MIchaelMainer commented:
@mikepizzo 
> adding deprecation to functions and methods based on binding/return type

Yes, it would be redundant as we can query the binding parameter at generation to discover whether it is deprecated and then apply /// comments based on the binding parameter deprecation to the method requests we generate.

baywet commented:
Hey @mikepizzo 
Thanks for looking into this. I don't think it's redundant by default, it really depends on who we want to take the "burden":
- vipr: responsible reading metadata and building a "context".
- typewriter: responsible for generating code from a specific context.
- customer's IDE: responsible for helping the dev when writing code.

To me option # 3 is subpart and depends too much on the language/IDE as some tell you when you're receiving a return type that's deprecated, some don't.
Option # 2 might lead to duplicated efforts by languages if we're not careful here.
Option # 1 looks like the best to me because not only it encapsulates the logic of deprecation propagation, each "component" is focused on it's purpose: typewriter writes code from a metadata context, vipr creates that metadata context from metadata files. On top of that your already have the code in place for the propagation.

mikepizzo commented:
Merged support for marking methods as deprecated based on binding/return type.  Also added additional tests for type definition and enum types.

Note that, for type definitions, we didn't previously scan for (any) vocabulary annotations, so I added that scan.  

Also, for type definitions we use the underlying type when generating properties of that type definition, so we lose the ability to apply any capability/deprecation information for properties based on a type definition.  However, if in the future we do add full support for retaining type definitions for properties, the deprecation will flow as part of that (and the test is already there and can be enabled).

mikepizzo commented:
Fixed.

---
In reply to: [482816409](https://github.com/microsoft/Vipr/pull/146#issuecomment-482816409) [](ancestors = 482816409)

mikepizzo commented:
@MIchaelMainer -- I believe I've addressed all comments and @baywet has signed off.  

Anything else needed before we can merge?  I'd like to get this merged so we finalize the support in the SDK (see related PR https://github.com/microsoftgraph/MSGraph-SDK-Code-Generator/pull/283, which is dependent upon this one).

MIchaelMainer reviewed (approved):
Thank you @mikepizzo for this work and bearing with us.
